tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml

```diff
@@ -1,6 +1,5 @@
 [package]
 name = "cargo_common_metadata"
 version = "0.1.0"
-publish = false
```

Comment:
Is removing this gonna be an issue when publishing Clippy?

Reply:
So these are there because of https://github.com/rust-lang/rust-clippy/pull/5614#discussion_r427308121

> Might be worth adding `publish = false` to avoid accidental publishing

So I'd rather keep it. You can keep the current tests working by disabling this check with checking for  the env var `__CLIPPY_INTERNAL_TESTS=true`. But then the test for `publish = false` won't work...

Maybe add a configuration option, where this can be configured to the lint. See https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md#adding-configuration-to-a-lint on how to do this.

Reply:
I would rather not add a lint configuration option considering this is only useful for testing Clippy itself and the Clippy configuration is a user-facing interface.

But I have a different suggestion, not adding a "description" and "license" makes it impossible to publish to Crates.io, so I can document this in a comment inside `Cargo.toml` to make sure nobody adds it on accident.

What do you think?

Reply:
The problem then still is, that for the `pass` tests it will always pass, just because they have the `publish = false`, not because they actually pass the tests.

You can hide the configuration from the user, by prefacing the lint name with `_`. Then it won't appear in the documentation.

Reply:
> The problem then still is, that for the `pass` tests it will always pass, just because they have the `publish = false`, not because they actually pass the tests.

I am afraid I don't understand what you are trying to tell me here, the `pass` tests don't always pass because they have `publish = false`, I just tried adding a `main.stderr` file in them and the test failed correctly.
Am I misunderstanding something here?

Reply:
So I believe I know what you meant now: the `pass` test fullfills all criterias to upload to Crates.io but doesn't have `publish = false`.

So I did implement your suggestion with the configuration option, but I can't make the test pick up the `clippy.toml` though, I'm assuming `ui-cargo` tests don't pick up on it.

Any help would be appreciated.

Reply:
So your check prevents this lint from triggering if `publish = false`. All `ui-cargo` tests have `publish = false` in their `Cargo.toml`. For tests which should trigger a lint (`fail` tests, with a `stderr` file), won't trigger this lint anymore with your check for `publish = false` and you see this in their tests failing because of a different `stderr`. But the same problem exists for tests that **shouldn't** trigger a lint (`pass` tests, no `stderr` file). Before this change they didn't produce a `stderr`, **because no lint triggered on the test**. But with your change the test behavior silently changed: The `pass` tests don't emit a `stderr` **because they have `publish = false` in them**.

So what you need to do is to add a `clippy.toml` file to every `ui-cargo` test with `_cargo_ignore_publish = true` in it, except the tests where you're testing your change.

---



> So I did implement your suggestion with the configuration option, but I can't make the test pick up the `clippy.toml` though, I'm assuming `ui-cargo` tests don't pick up on it.

The tests should pick up the clippy.toml file. But you have to add it to all tests, **except** the tests you add for your change.

---

I hope I explained this well enough now. This is understandably confusing.

Reply:
So I only used the workaround with `cargo-ignore-publish` on tests that actually need `publish = false`, but some tests don't need it because they don't fulfill the requirements to be published anyway, like a missing `author` or `description`.

In any case, I added `publish = false` back in, but how do I test `publish = true`? As far as I can see it's not a problem, because as mentioned above, it's not possible to publish this crate anyway.
> The tests should pick up the clippy.toml file.

As far as I tested, it doesn't pick it up, I solved it by adding `CLIPPY_CONF_DIR` to the testsuite, I'm open to any other suggestion.

Reply:
Everything seems to work just fine now.